### ISSUE TITLE
Parse a function (or lambda) only after all args are available

### ISSUE THREAD

**Issue by tavmem** [kevinlawler/kona#357]
Parsing a phrase entails considerable overhead.
If possible, it should be done only once.
In particular, a function (or lambda) should be opened for parsing only after all args are available.

@bakul provided the following examples:
```
Consider:
      f:{x+y}[1;2]
      f
    3

      f:{x+y}[1;2;3]
    valence error
    f:{x+y}[1;2;3]
      ^

      f:{x+y}[1;2;]
      f
    {x+y}[1;2;]

      f:{x+y}[1;]
      f
    {x+y}[1;]
      f 2
    3

      f:{x+y}[1;]
      f
    {x+y}[;2]
      f 3
    5

      f:{x+y+z}[1;;]
      f
    {x+y+z}[1;;]
      g:f[2]
      g
    {x+y+z}[1;;][2]
      g 3
    6

What these last few example reveal is that if an arg is
omitted, actual arg evaluation is left after sufficient
additional args have been supplied and inserted in the correct
place. Only then is the function actually evaluated. And this
is the point where it is discovered args do not match up.

In other words {...} is cracked open only once, after all the
args are available and only at that point can you discover
# of args don't match.
```
